internal/scheduler/reports.go

```diff
@@ -6,441 +6,314 @@ import (
+	for queue, qctx := range sctx.QueueSchedulingContexts {
+		for jobId, jctx := range qctx.SuccessfulJobSchedulingContexts {
+			jobSchedulingContextByJobId[jobId] = jctx
+		}
+		for jobId, jctx := range qctx.UnsuccessfulJobSchedulingContexts {
+			jobSchedulingContextByJobId[jobId] = jctx
 		}
+		queueSchedulingContextByQueue[queue] = qctx
 	}
-	return &schedulerobjects.QueueReport{
-		Report: report.String(),
-	}, nil
+	return queueSchedulingContextByQueue, jobSchedulingContextByJobId
 }
 
-func (repo *SchedulingReportsRepository) GetJobReport(ctx context.Context, jobId *schedulerobjects.JobId) (*schedulerobjects.JobReport, error) {
-	jobUuid, err := uuidFromUlidString(jobId.Id)
-	if err != nil {
-		return nil, err
-	}
-	report, ok := repo.GetJobSchedulingReport(jobUuid)
-	if !ok {
-		return nil, &armadaerrors.ErrNotFound{
-			Type:    "JobSchedulingReport",
-			Value:   jobId.Id,
-			Message: "this job has not been considered for scheduling recently",
-		}
-	}
-	return &schedulerobjects.JobReport{
-		Report: report.String(),
+func (repo *SchedulingContextRepository) GetSchedulingReport(_ context.Context, _ *types.Empty) (*schedulerobjects.SchedulingReport, error) {
+	return &schedulerobjects.SchedulingReport{
+		Report: repo.getSchedulingReportString(),
 	}, nil
 }
 
-func (repo *SchedulingReportsRepository) AddSchedulingRoundReport(report *SchedulingRoundReport) {
-	for queue, queueSchedulingRoundReport := range report.QueueSchedulingRoundReports {
-		repo.AddMany(queue, maps.Values(queueSchedulingRoundReport.SuccessfulJobSchedulingReports))
-		repo.AddMany(queue, maps.Values(queueSchedulingRoundReport.UnsuccessfulJobSchedulingReports))
+func (repo *SchedulingContextRepository) getSchedulingReportString() string {
+	var sb strings.Builder
+	w := tabwriter.NewWriter(&sb, 1, 1, 1, ' ', 0)
+	for _, executorId := range repo.sortedExecutorIds {
+		fmt.Fprintf(w, "%s:\n", executorId)
+		sctx := repo.mostRecentSchedulingContextByExecutor[executorId]
+		if sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent attempt:\n"))
+			fmt.Fprint(w, indent.String("\t\t", sctx.String()))
+		} else {
+			fmt.Fprint(w, indent.String("\t", "Most recent attempt: none\n"))
+		}
+		sctx = repo.mostRecentSuccessfulSchedulingContextByExecutor[executorId]
+		if sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent successful attempt:\n"))
+			fmt.Fprint(w, indent.String("\t\t", sctx.String()))
+		} else {
+			fmt.Fprint(w, indent.String("\t", "Most recent successful attempt: none\n"))
+		}
 	}
+	w.Flush()
+	return sb.String()
 }
 
-func (repo *SchedulingReportsRepository) AddMany(queueName string, reports []*JobSchedulingReport) {
-	for _, report := range reports {
-		repo.Add(queueName, report)
-	}
+func (repo *SchedulingContextRepository) GetQueueReport(_ context.Context, queue *schedulerobjects.Queue) (*schedulerobjects.QueueReport, error) {
+	queueName := strings.TrimSpace(queue.Name)
```

Comment:
why do we need to do this?  If we're getting this from an external system (e.g. it;'s coming over grpc) then I think we should do and validation/sanitisation in the service layer.

Reply:
Ok great- yeah we should separate storage from service but we can do that later.  I'd consider adding a comment to the function saying it's a user-facing RPC call, as I don't think this is obvious right now.

Reply:
Done